src/magnet_link.rs

```
@@ -86,6 +86,54 @@ impl MagnetLink {
   }
 }
 
+impl FromStr for MagnetLink {
+  type Err = Error;
+
+  fn from_str(text: &str) -> Result<Self, Self::Err> {
+    let mut link = None;
+    let url = Url::parse(&text).context(error::AnnounceUrlParse)?;
+
+    if url.scheme() != "magnet" {
+      return Err(Error::LinkParse {
+        text: "link must use the magnet:// scheme".into(),
+      });
+    }
+
+    for (k, v) in url.query_pairs() {
+      if k.as_ref() == "xt" {
+        if link.is_some() {
+          return Err(Error::LinkParse {
+            text: "link has multiple exact topics".into(),
+          });
+        }
+        let xt = v
+          .trim_start_matches("urn:bti")
```

If these aren't present it should return an error.